# Introduction

About two weeks ago, I was interviewing with one large telecommunications company for an analytics internship position, since the feedback is taking longer than expected, anxiety is killing me, so I think I should do something to keep my mind busy, `life is too short to be worried. LOL!`
This morning I was browsing through the internet for any controversial topic that I can build an analytics project on when I found [this](https://www.quora.com/Premier-League-Solve-an-argument-please-In-the-last-5-years-who-has-been-the-better-team-overall-Liverpool-or-Tottenham-Hotspur) Quora post about which team is better, overall, between Liverpool and Tottenham Hotspurs. Recently, I have not been watching football, meaning I am not very much inline with the current trends, but I do think this will be an interesting pursuit - to resolve this issue. Of course, I will need some of my hard-core fans of EPL to help me with the intuition as I work through this project.<br>

### Acknowledgements
Oh by the way, this approach to coming up with project ideas is primary inspired by Allen B. Downey, in this book **`Think Stats: Probability and Statistics for Programmers.`** In this book, he suggests an active learning approach where, if you're learning a concept/topic, particularly in statistics, come up with a topic that is of interest to you and then frame it into a statistical enquiry. These kinds of projects are more interesting if they are controversial and can allow for a back-and-forth of arguments, rather than something as obvious as the flat earth... HAHA LOL! I think I got you.<br>

Like many other informal internet debates or discussions, this problem is not `clearly` defined because a word like **"overall"** is opens a very large room for ambiguity. I do get what this Quora member is looking for, but I am going to try my best to make this question a **SMART** question. If you did something along the lines of project management or complex problem solving, then you should know what I mean, if you don't, the I guess, "Google is your friend." LOL!! So let's define the question...

# Problem Definition
Between Liverpool FC and Tottenham Hotspurs, which team has had better match statistics between the 2007-2008 and 2016-2017 seasons of the Barclays Premier League?<br>

Part of me really wants to provide a rigorous definition of this problem, discussing the trends that has been occurring between these two teams; but really, come on, this is the time for me to be having fun, so I am not going to stress over the McKinsey Way of solving problems. I love doing it, but not now please.

# Getting the Data
I have separate CSV files for each season from 07/08 season to 16/17, so I first have merge them, then subset the rows such that I only have games where either Liverpool or Tottenham is playing. This must be an interesting tasks. Let's see...<br>

In [1]:
import os, sys
import itertools
import pandas as pd

In [2]:
sys.path.append('../helper_funcs')

In [3]:
%load_ext autoreload
%autoreload 2

In [1]:
import helpers as hp

ModuleNotFoundError: No module named 'helpers'

In [ ]:
??hp.disjoint_columns

In [6]:
files = hp.create_dfs('07', '17')
files[0].head()

Div      Date       HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  E0  11/08/07    Aston Villa   Liverpool     1     2   A     0     1   A   
1  E0  11/08/07         Bolton   Newcastle     1     3   A     0     3   A   
2  E0  11/08/07          Derby  Portsmouth     2     2   D     1     1   D   
3  E0  11/08/07        Everton       Wigan     2     1   H     1     0   H   
4  E0  11/08/07  Middlesbrough   Blackburn     1     2   A     1     0   H   

    ...    BbMx>2.5  BbAv>2.5  BbMx<2.5  BbAv<2.5  BbAH  BbAHh  BbMxAHH  \
0   ...        2.30      2.14      1.70      1.63    26   0.50     1.95   
1   ...        2.33      2.10      1.75      1.65    26   0.00     1.83   
2   ...        2.35      2.13      1.90      1.63    25   0.00     2.26   
3   ...        2.20      2.00      1.83      1.73    26  -0.75     1.97   
4   ...        2.30      2.08      1.75      1.67    26   0.00     1.90   

   BbAvAHH  BbMxAHA  BbAvAHA  
0     1.85     2.07     1.95  
1     1.75     2.18     2.04  
2     2.09     1.76     1.71  
3     1.94     2.02     1.94  
4     1.77     2.10     1.97  

[5 rows x 71 columns]

## Next step
Make sure that there are consistent columns in all of the datasets and remove betting statistics

In [7]:
hp.disjoint_columns(files)

['SBA',
 'SBD',
 'SBH',
 'BSA',
 'BSD',
 'BSH',
 'GBA',
 'GBD',
 'GBH',
 'SJA',
 'SJD',
 'SJH']

All of the above columns are just are columns of the betting statistics, so I don't have to worry, I am going to delete them anyway.

In [8]:
files[0].columns

Index(['Div', 'Date', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG', 'FTR', 'HTHG',
       'HTAG', 'HTR', 'Referee', 'HS', 'AS', 'HST', 'AST', 'HF', 'AF', 'HC',
       'AC', 'HY', 'AY', 'HR', 'AR', 'B365H', 'B365D', 'B365A', 'BWH', 'BWD',
       'BWA', 'GBH', 'GBD', 'GBA', 'IWH', 'IWD', 'IWA', 'LBH', 'LBD', 'LBA',
       'SBH', 'SBD', 'SBA', 'WHH', 'WHD', 'WHA', 'SJH', 'SJD', 'SJA', 'VCH',
       'VCD', 'VCA', 'BSH', 'BSD', 'BSA', 'Bb1X2', 'BbMxH', 'BbAvH', 'BbMxD',
       'BbAvD', 'BbMxA', 'BbAvA', 'BbOU', 'BbMx>2.5', 'BbAv>2.5', 'BbMx<2.5',
       'BbAv<2.5', 'BbAH', 'BbAHh', 'BbMxAHH', 'BbAvAHH', 'BbMxAHA',
       'BbAvAHA'],
      dtype='object')

In [9]:
required_files = hp.subset_columns('Div', 'AR', files)

In [10]:
required_files[0].head()

Div      Date       HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  E0  11/08/07    Aston Villa   Liverpool     1     2   A     0     1   A   
1  E0  11/08/07         Bolton   Newcastle     1     3   A     0     3   A   
2  E0  11/08/07          Derby  Portsmouth     2     2   D     1     1   D   
3  E0  11/08/07        Everton       Wigan     2     1   H     1     0   H   
4  E0  11/08/07  Middlesbrough   Blackburn     1     2   A     1     0   H   

  ... HST  AST  HF  AF  HC  AC  HY  AY  HR  AR  
0 ...   6    7  18  11   4   2   4   2   0   0  
1 ...   9    5  15  16   4   3   1   1   0   0  
2 ...   5    6  14  17   6   6   1   2   0   0  
3 ...   8    4   8  13   6   2   0   0   0   0  
4 ...   6    4  16  16  13   3   3   4   0   0  

[5 rows x 23 columns]

## Next steps
1. Concatenate the dataframes
2. Identify and handle discrepancies in the data
    * Each team plays another, twice; once home and once away per season
    * Check distributions in other columns.
3. Only pull out data where Liverpool or Tottenham was playing

In [11]:
merged_data = pd.concat(required_files, axis=0)

In [12]:
required_files[0].shape

(380, 23)

In [13]:
merged_data.shape

(3801, 23)

In [14]:
merged_data.loc[380:,].head(3)

Div      Date     HomeTeam     AwayTeam  FTHG  FTAG  FTR  HTHG  HTAG  \
380  NaN       NaN          NaN          NaN   NaN   NaN  NaN   NaN   NaN   
0     E0  08/08/15  Bournemouth  Aston Villa   0.0   1.0    A   0.0   0.0   
1     E0  08/08/15      Chelsea      Swansea   2.0   2.0    D   2.0   1.0   

     HTR ...   HST   AST    HF    AF   HC   AC   HY   AY   HR   AR  
380  NaN ...   NaN   NaN   NaN   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
0      D ...   2.0   3.0  13.0  13.0  6.0  3.0  3.0  4.0  0.0  0.0  
1      H ...   3.0  10.0  15.0  16.0  4.0  8.0  1.0  3.0  1.0  0.0  

[3 rows x 23 columns]

Looks like the index column does not have unique values. That shouldn't be too much of a problem. I can reset it to default. I just want to make sure that, for each season, each team played another team only twice. We already know that, in each season, there are 38 games, hence, the reason it makes sense that we have 3800 games in 10 seasons. However, is this enough evidence? I don't think so.

### How am I going to do this??
Before merging, I think it would've been easy to create a season variable that I can now use work with seasons individually. Let's do that.
Now, we know that there 20 teams per season, each team plays 19 home games, which makes a total of 380 games per season.
I can find the mean of each team's number of home games and that should be *exactly* 19, and then find the average per seasons and it should also be *excatly* 19.<br>

Sounds like a complicated plan, but it's a plan with conclusive evidence

In [15]:
season_num = 0

for file in required_files:
    file['Season'] = season_num
    season_num += 1
    
required_files[0].head()

Div      Date       HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  E0  11/08/07    Aston Villa   Liverpool     1     2   A     0     1   A   
1  E0  11/08/07         Bolton   Newcastle     1     3   A     0     3   A   
2  E0  11/08/07          Derby  Portsmouth     2     2   D     1     1   D   
3  E0  11/08/07        Everton       Wigan     2     1   H     1     0   H   
4  E0  11/08/07  Middlesbrough   Blackburn     1     2   A     1     0   H   

    ...   AST  HF  AF  HC  AC  HY  AY  HR  AR  Season  
0   ...     7  18  11   4   2   4   2   0   0       0  
1   ...     5  15  16   4   3   1   1   0   0       0  
2   ...     6  14  17   6   6   1   2   0   0       0  
3   ...     4   8  13   6   2   0   0   0   0       0  
4   ...     4  16  16  13   3   3   4   0   0       0  

[5 rows x 24 columns]

In [16]:
merged_data = pd.concat(required_files, axis=0) # re-do the merging
merged_data.shape

(3801, 24)

In [17]:
merged_copy = merged_data.copy

In [18]:
type(merged_copy)

method

In [19]:
merged_data.groupby('Season').agg({'HomeTeam': 'size'})

HomeTeam
Season          
0            380
1            380
2            380
3            380
4            380
5            380
6            380
7            381
8            380
9            380

## How come we have 381 games in season 8?
Based on a quick Google search, there were 380 games played. I remember I saw a row - it was the 380th - with a lot of missing values, since season_agg_home shows means of 19 throughout, and no floating value anywhere, I suspect the additional game is due to those missing values, but first let me check if there are any redundancies/duplicates, if nothing happens, then I check if the reason is the missing values.

In [20]:
season8 = required_files[7]
season8.drop_duplicates(inplace=True)
season8.shape

(381, 24)

In [21]:
season8.dropna(inplace=True)

# Therefore
merged_data.dropna(inplace=True)

print(season8.shape)
print(merged_data.shape)

(380, 24)
(3800, 24)


### Fantastic!

In [22]:
season_agg_home = (merged_data.groupby(['Season', 'HomeTeam'])
 .agg({'HomeTeam': 'size'}))

In [23]:
(season_agg_home.groupby('Season')
 .agg({'HomeTeam': 'mean'})
 .rename(columns={'HomeTeam': 'HG_mean'})) # HG_mean for HomeGamesMean

HG_mean
Season         
0            19
1            19
2            19
3            19
4            19
5            19
6            19
7            19
8            19
9            19

### Now I am satisfied with the games
I can now subset the rows to get only the matches I am looking for.
Before that, however, let me check if other columns have weird values like outliers, I also think I want to change the column names, these abbreviations aren't readable.

In [ ]:
merged_data = (merged_data.rename(columns=
                                  {
                                      'FTHG': 'Final_HomeGoals',
                                      'FTAG': 'Final_AwayGoals',
                                      'FTR': 'Final_Result',
                                      'HTHG': 'Half_HomeGoals',
                                      'HTAG': 'Half_AwayGoals',
                                      'HTR': 'Half_Result',
                                      'HS': 'Home_Shots',
                                      'AS': 'Away_Shots',
                                      'HST': 'HomeShots_Target',
                                      'AST': 'AwayShots_Target',
                                      'HF': 'Home_Fouls',
                                      'AF': 'Away_Fouls',
                                      'HC': 'Home_Corner',
                                      'AC': 'Away_Corner',
                                      'HY': 'Home_Yellows',
                                      'AY': 'Away_Yellows',
                                      'HR': 'Home_Reds',
                                      'AR': 'Away_Reds'
                                  }))

### Are the names of teams all spelled correctly?
From the summary statistics below, I can see that there are only 35 teams which participated in the league in the last 10 seasons, that shouldn't require too much enerygy to go through and check the spelling

In [ ]:
team_names = list(itertools.chain
                  .from_iterable([merged_data['HomeTeam'], merged_data['AwayTeam']]))
team_names = set(team_names)

In [ ]:
team_names

### Team names are perfect!!

In [ ]:
merged_data.drop(labels='Div', axis=1, inplace=True)

In [ ]:
merged_data.describe(include='all').T

In [ ]:
merged_data.head()

In [ ]:
final_data = hp.subset_teams('Liverpool', 'Arsenal', merged_data)

In [ ]:
final_data.head()

In [ ]:
final_data.info()

In [ ]:
final_data['Final_Result'].astype('category')
final_data['Half_Result'].astype('category')

type(final_data['Final_Result'])

In [ ]:
final_data['Date'] = pd.to_datetime(final_data.Date)
type(final_data['Date']) # I don't know what this warning message means. I don't even think there has been a difference.

### Save the files for further analyses in other notebooks

In [ ]:
final_data.to_csv('../../Datasets/processed_data/liverpool_arsenal.csv')

In [ ]:
merged_data.to_csv('../../Datasets/processed_data/all_data.csv')

## Some things still missing

1. Confirm that it is always the case that when FTHG > FTAG then FTR == H. Do this for draws and Away wins
2. Same as the above with half-time results
3. Half-time score should always be equal to or less than full-time score
4. Shots on target must always be less that shots.

In [ ]:
add_scores = set(merged_data.apply(hp.check_scores, axis=1))
add_scores # Half-time and Full-time scores are fine

In [ ]:
odd_shots = merged_data.apply(hp.check_shots, axis=1)
print('Games with an unexpected number of shots: ', odd_shots.sum())

There is only 1 game with an odd number of shots. Maybe I shouldn't bother much. It was a game between Aston Villa and West Ham where Aston Villa had 5 shots but 8 shots on target. Since I am working with Liverpool and Tottenham, I guess I shouldn't worry about this.<br>

**Assumption:** I am therefore going to go with an assumption that variable shots, refers to all shots in the game; including shots-on-target. This is because - as we can see - only one game(a certain team in that game) had more shots-on-target than shots in 3800 games... What are the odds of shots and shots-on-target being indepedent of each other? Very few!